server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecider.java

```diff
@@ -104,4 +106,69 @@ public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode n
             return decision;
         }
     }
+
+    /**
+     * Returns a {@link Decision} whether the given shard can be forced to the
+     * given node in the event that the shard's source node is being vacated.
+     * This allows nodes using a vacate-type node shutdown (replace/vacate) to
+     * override certain deciders in the interest of moving the shard away from
+     * a node that *must* be removed.
+     *
+     * It defaults to returning "NO" and must be overridden by deciders that
+     * opt-in to having their other NO decisions overridden while vacating.
+     */
+    public Decision canForceDuringVacate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return Decision.NO;
+    }
+
+    /**
+     * Returns true if there are any node replacements ongoing in the cluster
+     */
+    public static boolean replacementOngoing(Metadata metadata) {
+        return metadata.nodeShutdowns().values().stream()
+            // TODO: change this to "REPLACE" type once that is available
+            .anyMatch(shutdown -> shutdown.getType().equals(SingleNodeShutdownMetadata.Type.REMOVE));
+    }
+
+    /**
+     * Returns true if there is a replacement currently ongoing from the source to the target node id
+     */
+    public static boolean replacementFromSourceToTarget(Metadata metadata, String sourceNodeId, String targetNodeId) {
+        if (replacementOngoing(metadata) == false) {
+            return false;
+        }
+        if (sourceNodeId == null || targetNodeId == null) {
+            return false;
+        }
+        // TODO: actually implement this once we have a replace type
+        return false;
+    }
+
+    /**
+     * Returns true if the given node id is the source (the replaced node) of an ongoing node replacement
+     */
+    public static boolean isReplacementSource(Metadata metadata, String nodeId) {
```

Comment:
I think these 3 helper methods should be moved somewhere less prominent, perhaps even as private methods on `NodeReplacementAllocationDecider`.